File: src/AudioPlayer/playback/PlayTimeHandler.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package AudioPlayer.playback;

import java.util.function.Function;
import javafx.util.Duration;
import util.async.executor.FxTimer;

/** Handler for handling events at specific point of song playback. */
public class PlayTimeHandler {
    private final FxTimer timer;
    private final Function<Duration,Duration> cal;

    /***
     * @param when_calculator calculates when the even should fire. The function
     * receives total song length and returns time at which the action executes
     * @param action action to execute
     */
    public PlayTimeHandler(Function<Duration,Duration> when_calculator, Runnable action) {
        timer = new FxTimer(Duration.INDEFINITE, 1, action);
        cal = when_calculator;
    }
    
    public void pause() {
        timer.pause();
    }
    
    public void unpause() {
        timer.unpause();
    }
    
    public void stop() {
        timer.stop();
    }
    
    public void restart(Duration total_time) {
        timer.start(cal.apply(total_time));
    }
    
    
    public static PlayTimeHandler at(Duration at, Runnable action) {
        return new PlayTimeHandler(total -> at, action);
    }
    
    public static PlayTimeHandler at(double perc, Runnable action) {
        return new PlayTimeHandler(total -> total.multiply(perc), action);
    }
    
}

File: src/player/main/sp/it/pl/audio/playback/PlayTimeHandler.kt
package sp.it.pl.audio.playback

import javafx.util.Duration
import sp.it.util.async.executor.FxTimer
import sp.it.util.async.executor.FxTimer.Companion.fxTimer

/** Handler for handling events at specific point of song playback. */
class PlayTimeHandler {

   private val whenCalculator: (Duration) -> Duration
   private val timer: FxTimer

   private constructor(whenCalculator: (Duration) -> Duration, action: () -> Unit) {
      this.whenCalculator = whenCalculator
      this.timer = fxTimer(Duration.INDEFINITE, 1, action)
   }

   fun pause() = timer.pause()

   fun unpause() = timer.unpause()

   fun stop() = timer.stop()

   fun restart(totalTime: Duration) = timer.start(whenCalculator(totalTime))

   companion object {

      /**
       * @param whenCalculator calculates when the event should fire. The function receives total song length and returns time at which the action executes
       * @param action action to execute
       */
      @JvmStatic
      fun at(whenCalculator: (Duration) -> Duration, action: () -> Unit) = PlayTimeHandler(whenCalculator, action)

   }

}